File: mobile/lib/core/money/money.dart
import 'currency.dart';

class Money {
  /// Amount represented strictly as integer minor units (e.g., $10.50 -> 1050 cents)
  final BigInt amountMinor;
  final Currency currency;

  const Money._({required this.amountMinor, required this.currency});

  factory Money.fromMinor(dynamic minor, Currency currency) {
    if (minor is BigInt) {
      return Money._(amountMinor: minor, currency: currency);
    } else if (minor is int) {
      return Money._(amountMinor: BigInt.from(minor), currency: currency);
    } else if (minor is String) {
      return Money._(amountMinor: BigInt.parse(minor), currency: currency);
    }
    throw ArgumentError('Invalid minor unit type: $minor');
  }

  factory Money.fromMajorString(String majorString, Currency currency) {
    final clean = majorString.replaceAll(',', '').trim();
    final parts = clean.split('.');
    final wholeStr = parts[0].isEmpty ? '0' : parts[0];
    final whole = BigInt.parse(wholeStr);

    var fractionStr = parts.length > 1 ? parts[1] : '';
    if (fractionStr.length > currency.decimals) {
      fractionStr = fractionStr.substring(0, currency.decimals);
    } else {
      fractionStr = fractionStr.padRight(currency.decimals, '0');
    }

    final fraction = fractionStr.isEmpty ? BigInt.zero : BigInt.parse(fractionStr);
    final multiplier = BigInt.from(10).pow(currency.decimals);

    final totalMinor = whole >= BigInt.zero
        ? (whole * multiplier) + fraction
        : (whole * multiplier) - fraction;

    return Money._(amountMinor: totalMinor, currency: currency);
  }

  factory Money.zero(Currency currency) {
    return Money._(amountMinor: BigInt.zero, currency: currency);
  }

  Money add(Money other) {
    _assertMatchingCurrency(other);
    return Money._(amountMinor: amountMinor + other.amountMinor, currency: currency);
  }

  Money subtract(Money other) {
    _assertMatchingCurrency(other);
    return Money._(amountMinor: amountMinor - other.amountMinor, currency: currency);
  }

  Money multiply(int scalar) {
    return Money._(amountMinor: amountMinor * BigInt.from(scalar), currency: currency);
  }

  /// Basis points calculation: 100 bps = 1.00%
  Money multiplyBasisPoints(int bps) {
    return Money._(
      amountMinor: (amountMinor * BigInt.from(bps)) ~/ BigInt.from(10000),
      currency: currency,
    );
  }

  bool isGreaterThan(Money other) {
    _assertMatchingCurrency(other);
    return amountMinor > other.amountMinor;
  }

  bool isLessThan(Money other) {
    _assertMatchingCurrency(other);
    return amountMinor < other.amountMinor;
  }

  bool get isZero => amountMinor == BigInt.zero;
  bool get isPositive => amountMinor > BigInt.zero;

  /// Returns standard major string (e.g., "1250.50")
  String toMajorString() {
    final isNegative = amountMinor < BigInt.zero;
    final absMinor = isNegative ? -amountMinor : amountMinor;
    final multiplier = BigInt.from(10).pow(currency.decimals);

    final whole = absMinor ~/ multiplier;
    final fraction = absMinor % multiplier;

    final fractionStr = fraction.toString().padLeft(currency.decimals, '0');
    return '${isNegative ? '-' : ''}$whole.$fractionStr';
  }

  /// Returns user-facing formatted string with symbol & comma separators (e.g. "$1,250.50" or "₦3,100,000.00")
  String formatFormatted({bool includeSymbol = true}) {
    final majorStr = toMajorString();
    final parts = majorStr.split('.');
    final whole = parts[0];
    final fraction = parts.length > 1 ? parts[1] : '00';

    // Insert commas
    final regExp = RegExp(r'\B(?=(\d{3})+(?!\d))');
    final formattedWhole = whole.replaceAll(regExp, ',');

    final prefix = includeSymbol ? currency.symbol : '';
    return '$prefix$formattedWhole.$fraction';
  }

  /// Minor units as string for BMONI endpoints expecting integer minor string (e.g. "1050")
  String toBmoniMinorString() => amountMinor.toString();

  void _assertMatchingCurrency(Money other) {
    if (currency != other.currency) {
      throw ArgumentError(
        'Currency mismatch: Cannot operate between ${currency.code} and ${other.currency.code}',
      );
    }
  }

  @override
  String toString() => formatFormatted();

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Money &&
          runtimeType == other.runtimeType &&
          amountMinor == other.amountMinor &&
          currency == other.currency;

  @override
  int get hashCode => amountMinor.hashCode ^ currency.hashCode;
}

File: backend/src/core/money.ts
/**
 * Central Money Abstraction
 * 
 * Guarantees:
 * 1. Financial amounts are ALWAYS stored and computed as integer minor units (e.g. cents, kobo, centavos).
 * 2. Zero floating-point arithmetic.
 * 3. Currency-aware operations (cannot accidentally add NGN to USD without explicit conversion).
 * 4. Safe formatting and serializing to/from BMONI API representations.
 */

export type SupportedCurrency = 'USD' | 'NGN' | 'MXN' | 'EUR' | 'CAD' | 'GBP';

// Standard decimals per currency (typically 2 for fiat/stablecoins USDB, CNGN, MEXe)
const CURRENCY_DECIMALS: Record<SupportedCurrency, number> = {
  USD: 2,
  NGN: 2,
  MXN: 2,
  EUR: 2,
  CAD: 2,
  GBP: 2,
};

export class Money {
  /** The amount in minor units (e.g., $10.50 -> 1050 cents) */
  readonly amountMinor: bigint;
  readonly currency: SupportedCurrency;

  private constructor(amountMinor: bigint | number, currency: SupportedCurrency) {
    this.amountMinor = BigInt(amountMinor);
    this.currency = currency;
  }

  /**
   * Create Money from integer minor units (e.g., 1050 for $10.50)
   */
  static fromMinor(amountMinor: bigint | number | string, currency: SupportedCurrency): Money {
    const minorBigInt = typeof amountMinor === 'bigint' ? amountMinor : BigInt(Math.round(Number(amountMinor)));
    return new Money(minorBigInt, currency);
  }

  /**
   * Create Money from major decimal string or number (e.g., "10.50" -> 1050)
   * Safely parses without floating point drift
   */
  static fromMajor(amountMajor: string | number, currency: SupportedCurrency): Money {
    const decimals = CURRENCY_DECIMALS[currency] ?? 2;
    const str = typeof amountMajor === 'number' ? amountMajor.toFixed(decimals) : amountMajor.trim();
    
    const parts = str.split('.');
    const whole = parts[0] ? BigInt(parts[0]) : 0n;
    let fractionStr = parts[1] ?? '';
    
    if (fractionStr.length > decimals) {
      fractionStr = fractionStr.slice(0, decimals);
    } else {
      fractionStr = fractionStr.padEnd(decimals, '0');
    }
    
    const fraction = BigInt(fractionStr);
    const multiplier = 10n ** BigInt(decimals);
    const totalMinor = whole >= 0n ? whole * multiplier + fraction : whole * multiplier - fraction;
    
    return new Money(totalMinor, currency);
  }

  static zero(currency: SupportedCurrency): Money {
    return new Money(0n, currency);
  }

  add(other: Money): Money {
    this.assertSameCurrency(other);
    return new Money(this.amountMinor + other.amountMinor, this.currency);
  }

  subtract(other: Money): Money {
    this.assertSameCurrency(other);
    return new Money(this.amountMinor - other.amountMinor, this.currency);
  }

  multiply(scalar: number | bigint): Money {
    const s = typeof scalar === 'bigint' ? scalar : BigInt(Math.round(scalar));
    return new Money(this.amountMinor * s, this.currency);
  }

  /**
   * Basis point percentage multiplication (e.g., 150 bps = 1.50%)
   * Useful for fee calculation with integer arithmetic
   */
  multiplyBasisPoints(bps: number | bigint): Money {
    const b = BigInt(bps);
    return new Money((this.amountMinor * b) / 10000n, this.currency);
  }

  isGreaterThan(other: Money): boolean {
    this.assertSameCurrency(other);
    return this.amountMinor > other.amountMinor;
  }

  isLessThan(other: Money): boolean {
    this.assertSameCurrency(other);
    return this.amountMinor < other.amountMinor;
  }

  isZero(): boolean {
    return this.amountMinor === 0n;
  }

  isPositive(): boolean {
    return this.amountMinor > 0n;
  }

  /**
   * Formats to standard major string (e.g., "10.50")
   */
  toMajorString(): string {
    const decimals = CURRENCY_DECIMALS[this.currency] ?? 2;
    const isNegative = this.amountMinor < 0n;
    const absolute = isNegative ? -this.amountMinor : this.amountMinor;
    
    const multiplier = 10n ** BigInt(decimals);
    const whole = absolute / multiplier;
    const fraction = absolute % multiplier;
    
    const fractionStr = fraction.toString().padStart(decimals, '0');
    return `${isNegative ? '-' : ''}${whole}.${fractionStr}`;
  }

  /**
   * BMONI format representation:
   * BMONI typically expects minor units as strings for certain endpoints (e.g. "1050")
   * or decimal strings for others.
   */
  toBmoniMinorString(): string {
    return this.amountMinor.toString();
  }

  toBmoniDecimalString(): string {
    return this.toMajorString();
  }

  toJSON() {
    return {
      amountMinor: this.amountMinor.toString(),
      currency: this.currency,
      formatted: this.toMajorString(),
    };
  }

  private assertSameCurrency(other: Money) {
    if (this.currency !== other.currency) {
      throw new Error(`Currency mismatch: cannot operate between ${this.currency} and ${other.currency}`);
    }
  }
}
